pkg/router/template/router.go

```diff
@@ -767,6 +773,11 @@ func (r *templateRouter) AddRoute(route *routeapi.Route) {
 
 	newConfig := r.createServiceAliasConfig(route, backendKey)
 
+	// We have to call the internal form of functions after this
+	// because we are holding the state lock.
+	r.lock.Lock()
+	defer r.lock.Unlock()
```

Comment:
It won't contend much... it's fed by the stream of events from the event queue and those are popped and handled one-by-one.  The other access is when the router state is written out, and that only happens periodically.

BTW the current model is really appalling, we do the following:
 - Pop event
 - Lock the structure
 - Update state
 - Unlock the structure
 - Call a rate-limited function to write state (default is no more often than 5s)

If it hasn't written state recently, it will:
 - Lock the structure
 - Write the conf file
 - Reload haproxy
 - Unlock the structure

On a system with lots of routes it can take 10+ seconds to reload haproxy... so we process one event and then reload... and do the same forever.

We have a card open to fix this.  We should only hold the state lock while the state is being written out.  But we need to not touch state again until the reload is complete... so we need to add another lock there.  And make sure that we don't block the event processing, just the event writing, while the reload happens.